Success markers: validate config

aeshky opened the pull request:
**Proposed changes**:
- Implemented https://github.com/RasaHQ/rasa/issues/9659

**Status (please check what you already did)**:
- [x] added some tests for the functionality
- [ ] updated the documentation
- [ ] updated the changelog (please check [changelog](https://github.com/RasaHQ/rasa/tree/main/changelog) for instructions)
- [x] reformat files using `black` (please check [Readme](https://github.com/RasaHQ/rasa#code-style) for instructions)

ancalita commented on rasa/core/evaluation/markers.py:
```diff
@@ -56,11 +56,11 @@ def from_file(cls, path: Text) -> Dict:
+
         except ParserError as e:
             e.filename = filename
             rasa.shared.utils.io.raise_warning(
```
Not directly related to this PR, but here I would actually suggest raising a `RasaException` since this would be a usage error (and to prevent a tsunami of Sentry errors because of invalid Yaml):
```suggestion
            raise RasaException(message=...) from e
```

ancalita replied:
I actually found a more useful subclass of `RasaException` [here.](https://github.com/RasaHQ/rasa/blob/0795099f5b237e648aed2e46397643f891da0378/rasa/shared/exceptions.py#L38)

ancalita commented on rasa/core/evaluation/markers.py:
```diff
@@ -56,11 +56,11 @@ def from_file(cls, path: Text) -> Dict:
 
-            return rasa.shared.utils.io.read_yaml(yaml)
+            config = rasa.shared.utils.io.read_yaml(yaml)
+            cls._validate_config(config)
```
Shouldn't `filename` argument be provided here too?

aeshky replied:
Well spotted.

ancalita commented on rasa/core/evaluation/markers.py:
```diff
@@ -93,3 +93,20 @@ def _merge(cls, config_a: Dict, config_b: Dict) -> Dict:
+        from jsonschema import ValidationError
+
+        try:
+            validate(config, MARKERS_SCHEMA)
```
You could call the static method `config_format_spec()` instead of directly using the constant, or you could delete the static method if you're not using it anywhere else.

aeshky replied:
Oh that was a mistake! I was calling the static method, but then moved stuff around while debugging my unit test. I don't call the static method anywhere else but I thought it might be useful to access the schema from the class itself? 🤔

ancalita commented on rasa/core/evaluation/markers.py:
```diff
@@ -93,3 +93,20 @@ def _merge(cls, config_a: Dict, config_b: Dict) -> Dict:
+            validate(config, MARKERS_SCHEMA)
+            return True
+        except ValidationError as e:
+            e.message += f". The file {filename} is invalid according to the schema."
```
```suggestion
            e.message += f". The file {filename} is invalid according to the json schema."
```
Are you considering listing the schema somewhere in the docs too? In that case it would be great to add the URL to the docs.

aeshky replied:
We don't yet have docs for the markers. I think I should create an issue to write the docs, and add an item to reference the docs in error messages.

aeshky replied:
Issue created: https://github.com/RasaHQ/rasa/issues/9748

ancalita replied:
👍🏼 Might be worth adding a tiny reminder in the issue to add the docs URL constant in error/ warning message(s) as applicable once these docs pages are available.

ancalita commented on rasa/shared/utils/schemas/markers.py:
```diff
@@ -0,0 +1,56 @@
+            "user_uttered": EVENT_SCHEMA,
+            "bot_uttered": EVENT_SCHEMA,
+        },
+    },
```
Should this also specify `minItems: 1` if you expect at least 1 dict in the array?

aeshky commented:
@usc-m and @alwx: Anca is reviewing this, but feel free to take a look and let me know if you have any feedback! :)

ancalita commented on tests/core/evaluation/test_markers_config.py:
```diff
@@ -104,17 +105,74 @@ def test_load_invalid_path():
+def test_config_missing_required_top_level_markers_label():
+    sample_json = {"marker": "no_restart"}
+    with pytest.raises(ValidationError):
+        MarkerConfig._validate_config(sample_json)
```
I would recommend using the public method `from_yaml` directly in each test, since `_validate_config` should not be accessed outside the `MarkerConfig` class.

aeshky replied:
I changed `_validate_config` to a public method. I think it works conceptually as a public method. What do you think?

ancalita replied:
If you expect this method to be accessed in other places in the codebase, then that's fine.

ancalita commented on tests/core/evaluation/test_markers_config.py:
```diff
@@ -104,17 +105,74 @@ def test_load_invalid_path():
+
+
+def test_config_invalid_event():
+    sample_json = {
```
if you change to using `from_yaml`, you'll also have to change to using yaml and that will make the tests more readable in my opinion.

aeshky replied:
I find Json easier to work with 🙈 yaml formatting drives me crazy.

ancalita replied:
Pro-tip: You could use `textwrap.dedent(<yaml-content>)` to get rid of some of the whitespace related issues.

aeshky replied:
Cool!

ancalita reviewed (changes requested):
Speedy work 💯 Left a few suggestions.

aeshky commented on data/test_markers/config_invalid.yml:
```diff
@@ -0,0 +1,7 @@
+markers:
+- marker: incorrectly_indented_yaml
+  conditions:
```
The label used to be plural but I changed it to singular. This file is meant to have an invalid yaml syntax anyway, but I'll change it to singular since I'm using it to catching a different error.

aeshky commented:
Thanks for the thorough review @ancalita 💯

ancalita commented on rasa/core/evaluation/markers.py:
```diff
@@ -56,21 +56,18 @@ def from_file(cls, path: Text) -> Dict:
-            rasa.shared.utils.io.raise_warning(
-                message=f"The file {filename} could not be loaded "
+            raise YamlSyntaxException(
+                f"The file {filename} could not be loaded "
```
I actually think the constructor of `YamlSyntaxEception` only accepts `filename` and `e` arguments, and not message, since this exception class has a custom `__str__` method. Otheriwse this string you've added as message will be considered the `filename` in its constructor and the message could end up being confusing.

ancalita reviewed (commented):
Looks good, a few more details to iron out 🎉 
I still advocate for testing `from_yaml` with invalid config file according to the json schema to make sure it raises the `ValidationError`.

aeshky commented on rasa/shared/utils/schemas/markers.py:
```diff
@@ -0,0 +1,57 @@
+            "user_uttered": EVENT_SCHEMA,
+            "bot_uttered": EVENT_SCHEMA,
+        },
+        "minItems": 1,
```
You're right. I added a unit test for it.

aeshky commented:
Thanks Anca! I've addressed your comments.
>I still advocate for testing from_yaml with invalid config file according to the json schema to make sure it raises the ValidationError.

I switched to using `from_yaml` in the tests 👍

aeshky commented:
Any recommendations for the failing check? 
> codeclimate/diff-coverage — 0% (50% threshold)

ancalita commented on rasa/core/evaluation/markers.py:
```diff
@@ -56,21 +56,13 @@ def from_file(cls, path: Text) -> Dict:
-                category=UserWarning,
-            )
-            raise e
+            raise YamlSyntaxException(filename)
```
Optional:
```suggestion
            raise YamlSyntaxException(filename, e)
```

ancalita reviewed (approved):
🎉

ancalita commented:
> Any recommendations for the failing check?
> 
> > codeclimate/diff-coverage — 0% (50% threshold)

@mprazz do you know what is the issue with this failing check and how to remedy it? I've had a quick look at Details and it seemed the repo is rated at [18% test coverage](https://codeclimate.com/github/RasaHQ/rasa/trends/test_coverage_total)?

mprazz commented:
> > Any recommendations for the failing check?
> > > codeclimate/diff-coverage — 0% (50% threshold)
> 
> @mprazz do you know what is the issue with this failing check and how to remedy it? I've had a quick look at Details and it seemed the repo is rated at [18% test coverage](https://codeclimate.com/github/RasaHQ/rasa/trends/test_coverage_total)?

@ancalita So this specific check is regarding coverage for new code (i.e. code introduced in this PR) and it's failing because there is no text coverage for the new code -- as for the overall coverage, indeed is it rather low, most likely because we need to edit the exclusion patterns in the codeclimate config file -- I can do it but I need some help making sure I have the right patterns, basically that I am only excluding the parts that we don't care about. Maybe someone from your squad can help me out on this, or edit the config file yourselves? :pray:

ancalita commented:
> it's failing because there is no text coverage for the new code

@mprazz that's quite strange, there are plenty of tests actually added to test the new code 🤔 
I'll nudge my squad to reply to you on the config file patterns.

mprazz commented:
> > it's failing because there is no text coverage for the new code
> 
> @mprazz that's quite strange, there are plenty of tests actually added to test the new code thinking I'll nudge my squad to reply to you on the config file patterns.

That is indeed strange!! I can look into that with whoever has a bit of time from your squad as well -- but if you want to check it as well, all individual coverage reports are added to the workflow artifacts before being uploaded to codeclimate, so maybe you'll be able to spot some issue there (perhaps on the `coverage` command I used?). On my end I can check the uploaded report on codeclimate, since I'm admin there. Happy to collab on this :blush: :pray:

aeshky commented:
@mprazz and @ancalita I wonder if it's complaining that we're not testing [rasa/shared/utils/schemas/markers.py](https://github.com/RasaHQ/rasa/pull/9746/files#diff-eb40b8559b01f82a6a90e887aa403bac23896bf80014e4cc4cd5fac56fe03889) which is a json schema file but with a `.py` extension? The schema is tested indirectly via the class. Or maybe it's all the `.yaml` files added in `data/test_markers` (used as input to tests)?

aeshky commented:
@mprazz It passed. Did you change anything :) ?

> codeclimate All good!

mprazz commented:
> @mprazz It passed. Did you change anything :) ?
> 
> > codeclimate All good!

I changed nothing! :sweat_smile:
I made a PR for the exclusion patterns (#9756 ), and will check what's up on this PR a bit later when I have some help :pray:

aeshky commented:
Haha that's weird. All I did was update the branch. Let me know how to view the report from codeclimate if that's helpful.
